File: 3097-shortest-subarray-with-or-at-least-k-ii/3097-shortest-subarray-with-or-at-least-k-ii.java
class Solution {
    public int minimumSubarrayLength(int[] nums, int k) {
        int n = nums.length, minLen = Integer.MAX_VALUE, currOr = 0;
        var bitCount = new int[32];

        for(int l=0, r=0; r<n; r++){
            // base case
            if(nums[r] >= k) return 1;

            // update cumulative OR
            currOr = addNumToOr(nums[r], currOr, bitCount);
            
            // shift left pointer till OR stops matching
            while(currOr >= k){
                minLen = Math.min(minLen, r - l + 1);
                currOr = removeNumFromOr(nums[l++], currOr, bitCount);
            }
        }
        
        if(minLen == Integer.MAX_VALUE) return -1;
        return minLen;
    }

    private int removeNumFromOr(int num, int currOr, int[] bitCount){
        for(var b=0; b<32 && num > 0; b++){
            bitCount[b] -= num & 1;
            num >>= 1;

            if(bitCount[b] == 0)
                currOr &= (~(1 << b));
        }

        return currOr;
    }

    private int addNumToOr(int num, int currOr, int[] bitCount){
        currOr |= num;
        for(var b=0; b<32 && num > 0; b++){
            bitCount[b] += num & 1;
            num >>= 1;
        }

        return currOr;
    }
}

File: 3097-shortest-subarray-with-or-at-least-k-ii/3097-shortest-subarray-with-or-at-least-k-ii.py
class Solution:
    def minimumSubarrayLength(self, nums: List[int], k: int) -> int:
        n, minLen, currOr = len(nums), len(nums)+1, 0
        bitCount = [0] * 32

        l = 0
        for r in range(n):
            # base case
            if nums[r] >= k: return 1

            # update cumulative OR
            currOr = self.addNumToOr(nums[r], currOr, bitCount)
            
            # shift left pointer till OR stops matching
            while currOr >= k:
                minLen = min(minLen, r - l + 1)
                currOr = self.removeNumFromOr(nums[l], currOr, bitCount)
                l += 1

        if minLen == n+1: return -1
        return minLen

    def removeNumFromOr(self, num, currOr, bitCount):
        for b in range(32):
            if num == 0: break
            bitCount[b] -= num & 1
            num >>= 1

            if bitCount[b] == 0:
                currOr &= (~(1 << b))

        return currOr

    def addNumToOr(self, num, currOr, bitCount):
        currOr |= num
        for b in range(32):
            if num == 0: break
            bitCount[b] += num & 1
            num >>= 1

        return currOr